File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> numbers) {
    	vector<int> answer(numbers.size(), -1);
	vector<int> Index(numbers.size(), -1);
	int Max = -1;
	int MaxIndex = 0;
    	for (int i = numbers.size() - 1; i >= 0; i--)
	{
		if (numbers[i] >= Max)
		{
			Max = numbers[i];
			Index[i] = i;
			MaxIndex = i;
		}
		else
		{
			for (int j = i + 1; j <= MaxIndex;)
			{
				if (numbers[i] < numbers[j])
				{
					answer[i] = numbers[j];
					Index[i] = j;
					break;
				}
				else
				{
					j = Index[j];
				}
			}
		}
	}
    return answer;
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public int[] solution(int[] numbers) {
        int[] answer = new int[numbers.Length];
        Stack<int> stk = new Stack<int>();
        for(int i = 0; i<numbers.Length;i++)
        {
            answer[i] = -1;
            int n = numbers[i];
            while(stk.Count > 0)
            {
                if(numbers[stk.Peek()] < n)
                {
                    answer[stk.Pop()] = n;
                }
                else break;
            }
            stk.Push(i);
        }
        
        return answer;
    }
}